src/LiftType.hs

```diff
@@ -67,38 +68,42 @@ liftType =
         case tr of
             Con tyCon ->
                 mk tyCon
-            App trA trB ->
-                AppT (go trA) (go trB)
             Fun trA trB ->
                 ConT ''(->) `AppT` go trA `AppT` go trB
+            App trA trB ->
+                AppT (go trA) (go trB)
 
     mk :: TyCon -> Type
     mk tyCon =
         let
             tcName =
                 tyConName tyCon
+            trySymbol = case tcName of
+              '"' : cs -> Just $ LitT (StrTyLit (zipWith const cs (drop 1 cs)))
+              _ -> Nothing
+            tryTicked = case tcName of
+              '\'' : dcName -> Just (PromotedT name)
+                where
+                  nameBase =
+                      mkOccName dcName
+
+                  flavor =
+                      NameG
+                          DataName
+                          (mkPkgName $ tyConPackage tyCon)
+                          (mkModName $ tyConModule tyCon)
+                  name =
+                      Name
+                          nameBase
+                          flavor
+              _ -> Nothing
+            tryNat = LitT . NumTyLit <$> readMaybe tcName
         in
-            if hasTick tcName
-            then
-                let
-                    nameBase =
-                        mkOccName (drop 1 tcName)
-                    flavor =
-                        NameG
-                            DataName
-                            (mkPkgName $ tyConPackage tyCon)
-                            (mkModName $ tyConModule tyCon)
-                    name =
-                        Name
-                            nameBase
-                            flavor
-                in
-                    PromotedT name
-            else if hasDigit tcName then
-                LitT (NumTyLit (read tcName))
-            else if hasQuote tcName then
-                LitT (StrTyLit (stripQuotes tcName))
-            else
+            if
+              | Just tc <- tryTicked -> tc
+              | Just tc <- trySymbol -> tc
+              | Just tc <- tryNat -> tc
+              | otherwise ->
```

Comment:
I like the `try` approach quite a bit, but I think it might be better if we used `asum` instead of `MultiWayIf`:

```haskell
fromMaybe otherwise_ $ asum [tryTicked, trySymbol, tryNat]
```

Reply:
I considered `Alternative` too. It doesn't make much difference to me.